archive/finder.go

```diff
@@ -11,83 +11,112 @@ import (
+	if pathField == "" {
+		panic("missing pathField argument")
+	}
+	// Create a type alias called "zfile" that the client will
+	// understand as the pathname of a zng file, e.g., available
+	// as a new space via zqd.
+	return func(opt *findOptions) error {
+		typ, err := opt.zctx.LookupTypeAlias("zfile", zng.TypeString)
+		if err != nil {
+			return err
+		}
+		pathCol := []zng.Column{{pathField, typ}}
+		opt.addPath = func(ark *Archive, si SpanInfo, rec *zng.Record) (*zng.Record, error) {
+			var path string
+			if absolutePath {
+				path = si.LogID.Path(ark)
+			} else {
+				path = string(si.LogID)
+			}
+			val := zng.Value{pathCol[0].Type, zng.EncodeString(path)}
+			return opt.zctx.AddColumns(rec, pathCol, []zng.Value{val})
+		}
+		return nil
+	}
+}
+
 // Find descends a directory hierarchy looking for index files to search and
-// for each such index, it looks up the given pattern.  If the pattern matches,
+// for each such index, it executes the given query. If the pattern matches,
 // a zng.Record of that row in the index is streamed to the hits channel.
 // If multiple rows match, they are streamed in the order encountered in the index.
 // Multiple records can match for a multi-key search where one or more keys are
 // unspecified, implying a "don't care" condition for the unspecified sub-key(s).
-// If skipMissing is true, then log files that do not have the indicated index
+// If the SkipMissing option is used, then log files that do not have the indicated index
 // are silently skipped; otherwise, an error is returned for the missing index file
-// and the search is terminated.  If pathField is set, then a column is added
+// and the search is terminated.  If the AddPath option is used, then a column is added
 // to each returned zng.Record with the path of the log file relating that hit
 // where the column name is given by the pathField argument.  If the passed-in
 // ctx is canceled, the search will terminate with error context.Canceled.
 // XXX We currently allow only one multi-key pattern at a time though it might
 // be more efficient at large scale to allow multipe patterns that
 // are effectively OR-ed together so that there is locality of
 // access to the zdx files.
-func Find(ctx context.Context, ark *Archive, indexName string, pattern []string, hits chan<- *zng.Record, pathField string, skipMissing bool) error {
-	//XXX this can be parallelized fairly easily since the search results are
-	// share the same type context allowing the zng.Records to easily comprise
-	// a single stream
-	zctx := resolver.NewContext()
-	var pathCol []zng.Column
-	if pathField != "" {
-		// Create a type alias called "zfile" that the client will
-		// understand as the pathname of a zng file, e.g., available
-		// as a new space via zqd.
-		typ, err := zctx.LookupTypeAlias("zfile", zng.TypeString)
-		if err != nil {
+func Find(ctx context.Context, ark *Archive, query IndexQuery, hits chan<- *zng.Record, opts ...FindOption) error {
```

Comment:
This is the first usage of functional options in the zq repo, so an fyi on my rationale here:
- addPath needs to happen in the search loop so it has access to the originating log file of the record.
- addPath is almost just a transform on the output records, but it needs a small amount of setup (the `LookupTypeAlias` call), and it needs the `zctx` at both setup time & while it operates.
- addPath takes two parameters that are only relevant to itself, not the overall find/search operation.
- addPath isn't used for every find call.

The above led me to consider how to make addPath a kind of optional transformation that gets called within Find, and the zctx usage for its setup means it needed two parts. Overall, that seemed like a good match for the functional option technique.

(fyi, I think the `driver.Compile` related functions & arguments would also work well as functional options.)